SU2_CFD/src/solvers/CGradientSmoothingSolver.cpp

```diff
@@ -0,0 +1,1168 @@
+  unsigned int iDim;
+  for (iPoint = 0; iPoint < nPoint; iPoint++) {
+    for (iDim = 0; iDim < nDim; iDim++) {
+      geometry->SetSensitivity(iPoint,iDim, vector[iPoint*nDim+iDim]);
+    }
+  }
+}
+
+void CGradientSmoothingSolver::ReadVector2Geometry(CGeometry *geometry, CConfig *config, CSysVector<su2matvecscalar> &vector) {
+  unsigned long iPoint;
+  unsigned int iDim;
+  for (iPoint = 0; iPoint < nPoint; iPoint++) {
+    for (iDim = 0; iDim < nDim; iDim++) {
+      vector[iPoint*nDim+iDim] = SU2_TYPE::GetValue(geometry->GetSensitivity(iPoint,iDim));
+    }
+  }
+}
+
+void CGradientSmoothingSolver::Set_VertexEliminationSchedule(CGeometry *geometry, CConfig *config) {
+
+  /*--- Store global point indices of essential BC markers. ---*/
+  vector<unsigned long> myPoints;
+  unsigned long iPoint, iVertex;
+
+  /*--- get global indexes of Dirichlet boundaries ---*/
+
+  if (config->GetSmoothOnSurface()) {
+
+    /*--- Find Marker_DV if this node has any ---*/
+    unsigned long dvMarker=NOT_AVAILABLE;
+    for (unsigned short iMarker = 0; iMarker < config->GetnMarker_All(); iMarker++) {
+      if (config->GetMarker_All_DV(iMarker) == YES) {
+        dvMarker = iMarker;
+      }
+    }
+
+    /*--- Surface case:
+     * Fix design boundary border if Dirichlet condition is set and the current rank holds part of it. ---*/
+    if ( config->GetDirichletSurfaceBound() && dvMarker!=NOT_AVAILABLE) {
+      for (iVertex = 0; iVertex < geometry->nVertex[dvMarker]; iVertex++) {
+        /*--- Get node index ---*/
+        iPoint = geometry->vertex[dvMarker][iVertex]->GetNode();
+        if ( nodes->IsBoundaryPoint(iPoint) ) {
+          myPoints.push_back(geometry->nodes->GetGlobalIndex(iPoint));
+        }
+      }
+    }
+
+  } else {
+
+    /*--- Volume case:
+     * All boundaries where Dirichlet conditions are set. ---*/
+    vector<unsigned short> markers;
+    for (unsigned short iMarker = 0; iMarker < config->GetnMarker_All(); iMarker++) {
+      if (config->GetMarker_All_SobolevBC(iMarker) == YES) {
+        markers.push_back(iMarker);
+      }
+    }
+    for (auto iMarker : markers) {
+      for (iVertex = 0ul; iVertex < geometry->nVertex[iMarker]; iVertex++) {
+        auto iPoint = geometry->vertex[iMarker][iVertex]->GetNode();
+        myPoints.push_back(geometry->nodes->GetGlobalIndex(iPoint));
+      }
+    }
+  }
+
+  /*--- communicate the boundary points ---*/
+
+  const unordered_set<unsigned long> markerPoints(myPoints.begin(), myPoints.end());
+
+  vector<unsigned long> numPoints(size);
+  unsigned long num = myPoints.size();
+  SU2_MPI::Allgather(&num, 1, MPI_UNSIGNED_LONG, numPoints.data(), 1, MPI_UNSIGNED_LONG, SU2_MPI::GetComm());
+
+  /*--- Global to local map for the halo points of the rank (not covered by the CGeometry map). ---*/
+  unordered_map<unsigned long, unsigned long> Global2Local;
+  for (auto iPoint = nPointDomain; iPoint < nPoint; ++iPoint) {
+    Global2Local[geometry->nodes->GetGlobalIndex(iPoint)] = iPoint;
```

Comment:
We need a way to avoid duplicating this function, or parts of it at least.

Reply:
The new CFEASolverBase class handles this and the element container for all solvers, which use the internal finite elements (e.g. FEA and Mesh solver).